File: mobile/src/main/java/com/alexstyl/specialdates/upcoming/ui/UpcomingViewHolderFactory.java
package com.alexstyl.specialdates.upcoming.ui;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.alexstyl.specialdates.R;
import com.alexstyl.specialdates.images.ImageLoader;
import com.alexstyl.specialdates.upcoming.UpcomingRowViewType;
import com.alexstyl.specialdates.upcoming.view.UpcomingEventsView;
import com.novoda.notils.caster.Views;
import com.novoda.notils.exception.DeveloperError;

public final class UpcomingViewHolderFactory {

    private final LayoutInflater layoutInflater;
    private final ImageLoader imageLoader;

    public UpcomingViewHolderFactory(LayoutInflater layoutInflater, ImageLoader imageLoader) {
        this.layoutInflater = layoutInflater;
        this.imageLoader = imageLoader;
    }

    public UpcomingRowViewHolder createFor(int viewType, ViewGroup parent) {
        if (viewType == UpcomingRowViewType.UPCOMING_EVENTS) {
            View view = layoutInflater.inflate(R.layout.row_upcoming_day, parent, false);
            TextView titleView = Views.findById(view, R.id.upcoming_events_title);
            UpcomingEventsView eventsView = Views.findById(view, R.id.upcoming_events_view);
            return new UpcomingEventsViewHolder(view, titleView, eventsView, imageLoader);
        } else if (viewType == UpcomingRowViewType.MONTH_HEADER) {
            View view = layoutInflater.inflate(R.layout.row_header_month, parent, false);
            TextView monthView = (TextView) view.findViewById(android.R.id.text1);
            return new MonthHeaderViewHolder(view, monthView);
        } else {
            throw new DeveloperError("Invalid viewType " + viewType);
        }
    }
}

File: android_mobile/src/main/java/com/alexstyl/specialdates/upcoming/UpcomingViewHolderFactory.kt
package com.alexstyl.specialdates.upcoming

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import com.alexstyl.specialdates.R
import com.alexstyl.specialdates.images.ImageLoader
import com.alexstyl.specialdates.ui.widget.ColorImageView
import com.novoda.notils.exception.DeveloperError

internal class UpcomingViewHolderFactory(private val layoutInflater: LayoutInflater, private val imageLoader: ImageLoader) {

    fun createFor(@UpcomingRowViewType viewType: Int, parent: ViewGroup): UpcomingRowViewHolder<*> {
        when (viewType) {
            UpcomingRowViewType.DATE_HEADER -> {
                val view = layoutInflater.inflate(R.layout.row_upcoming_events_date_header, parent, false)
                val dateView = view.findViewById<TextView>(R.id.upcoming_event_date_header)
                return DateHeaderViewHolder(view, dateView)
            }
            UpcomingRowViewType.BANKHOLIDAY -> {
                val view = layoutInflater.inflate(R.layout.row_upcoming_events_bankholiday, parent, false)
                val holidayName = view.findViewById<TextView>(R.id.row_upcoming_bankholiday)
                return BankholidayViewHolder(view, holidayName)
            }
            UpcomingRowViewType.NAMEDAY_CARD -> {
                val view = layoutInflater.inflate(R.layout.row_upcoming_events_nameday, parent, false)
                val namedays = view.findViewById<TextView>(R.id.row_upcoming_namedays)
                return NamedaysViewHolder(view, namedays)
            }
            UpcomingRowViewType.CONTACT_EVENT -> {
                val view = layoutInflater.inflate(R.layout.row_upcoming_events_contact_event, parent, false)
                val avatarView = view.findViewById<ColorImageView>(R.id.row_upcoming_event_contact_image)
                val contactName = view.findViewById<TextView>(R.id.row_upcoming_event_contact_name)
                val eventLabel = view.findViewById<TextView>(R.id.row_upcoming_event_contact_event)
                return ContactEventViewHolder(view, avatarView, contactName, eventLabel, imageLoader)
            }
            else -> {
                throw DeveloperError("Unhandled viewType [$viewType]")
            }
        }
    }
}
